### PR TITLE
Add universal flashing script

### PR DESCRIPTION
**Pull request by unbel13ver**
The 'flash.sh' script makes image flashing easy and hassle-free, both for compressed and uncompressed images.

## Description of changes
These changes should fix the recent issue with ZFS, when the device does not boot after the storage media was reflashed.

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [ ] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [x] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

## Testing

Build an image, flash it to the media using `flash.sh` script and boot the device.

### CODE REVIEW

**Review by leivos-unikie — COMMENTED** [review]
@milva-unikie noticed that she had to replace `#!/bin/bash` with `#!/run/current-system/sw/bin/bash` in her nixos environment. 

How about using `#!/usr/bin/env bash` ?

How universal should the script be? Is it meant that user modifies these
```
OUTPUT_DIR="result"
UNCOMPRESSED_NAME="nixos.img"
COMPRESSED_NAME="disk1.raw.zst"
```

Now it works fine for flashing lenovo-x1 image (result/disk1.raw.zst) or icicle-kit image (result/nixos.img). For using the script for other targets:
- lenovo-x1 installer (e.g. result/iso/ghaf-24.11.20240802.c488d21-x86_64-linux.iso)
- orin agx/nx (e.g. /result/sd-image/nixos-sd-image-24.11.20240802.c488d21-aarch64-linux.img.zst)
one has to modify.

**Comment by milva-unikie** [comment]
> Do we need to change the [ghaf-installer.sh](https://github.com/tiiuae/ghaf/blob/main/packages/installer/ghaf-installer.sh) script also?

Lenovo-X1 images installed with the installer get stuck at boot, so it definitely should also be fixed. 

> Now it works fine for flashing lenovo-x1 image (result/disk1.raw.zst) or icicle-kit image (result/nixos.img). For using the script for other targets:
> 
>     * lenovo-x1 installer (e.g. result/iso/ghaf-24.11.20240802.c488d21-x86_64-linux.iso)
> 
>     * orin agx/nx (e.g. /result/sd-image/nixos-sd-image-24.11.20240802.c488d21-aarch64-linux.img.zst)
>       one has to modify.

Also to add to this, builds made in Himalia and Jenkins produce their own output packages with different names. The latest Lenovo-X1 Himalia-image for example is: `d1cnkzqcpd0rp92ifkgdlhszakjb8pjx-ghaf-host-disko-images-himalia-621/d1cnkzqcpd0rp92ifkgdlhszakjb8pjx-ghaf-host-disko-images/disk1.raw.zst`

File: packages/flash/flash.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+if [ ! -e "$DEVICE" ]; then
+    echo "No such block device: ${DEVICE}"
+    exit 1
+fi
+
+if [ ! -e "$FILENAME" ]; then
+    echo "No such file: ${FILENAME}"
+    exit 1
+fi
+
```
[the remark this answers is not in the document]
Comment (unbel13ver) [reply packages/flash/flash.sh:35]: With ZFS, as long as the image file itself fits a storage media, everything works just fine. I successfully tested the image for Lenovo X1 using 32Gb microSD card.

File: packages/flash/default.nix
[the remark this answers is not in the document]
Comment (unbel13ver) [reply packages/flash/default.nix:1]: Done.

**Comment by unbel13ver** [comment]
> Since we have different paths depending on the target, maybe it is better to generate a flash script with the build, like the Orin: https://github.com/tiiuae/ghaf/blob/main/lib/mk-flash-script/default.nix

I initially wanted to keep this flashing script as a separate independent entity and do not contaminate target files. So I've added the possibility to pass the image file name as an argument.

**Comment by leivos-unikie** [comment]
For the Orin images on vedenemo.dev one just has to first uncompress the *.tar.xz and then the flash script is applicable.

File: packages/flash/flash.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    sync; sudo umount -q "$DEVICE"* || true;
+    # Wipe first 10MiB of disk
+    sudo dd if=/dev/zero of="$DEVICE" bs="$SECTOR" count="$MIB_TO_SECTORS" conv=fsync status=none
+    # Wipe last 10MiB of disk
+    sudo dd if=/dev/zero of="$DEVICE" bs="$SECTOR" count="$MIB_TO_SECTORS" seek="$((SECTORS - MIB_TO_SECTORS))" conv=fsync status=none
+    echo "Flashing..."
+}
+
+# Ask for sudo
+sudo -v
```
[the remark this answers is not in the document]
Comment (unbel13ver) [reply packages/flash/flash.sh]: This is a very good point, thanks. I always forget about this since all systems I have ever used used `sudo`=)
Addressed.

**Comment by unbel13ver** [comment]
> For the Orin images on vedenemo.dev one just has to first uncompress the *.tar.xz and then the flash script is applicable.

This is expected since the script is used in the "Build and Run" section of the documentation. The `vedenemo` images are mentioned in "Release Notes", though. So I can add `*.tar.xz` support to the script, if you think it is needed.

**Comment by leivos-unikie** [comment]
> > For the Orin images on vedenemo.dev one just has to first uncompress the *.tar.xz and then the flash script is applicable.
> 
> This is expected since the script is used in the "Build and Run" section of the documentation. The `vedenemo` images are mentioned in "Release Notes", though. So I can add `*.tar.xz` support to the script, if you think it is needed.

I think it's enough that flash script covers the 'run and build' case. No need to serve everything.

**Review by leivos-unikie — APPROVED** [review]

File: packages/flash/flash.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    echo "Wiping filesystem..."
+    # Set sector size to 512 bytes
+    SECTOR=512
+    # 10 MiB in 512-byte sectors
+    MIB_TO_SECTORS=20480
+    # Disk size in 512-byte sectors
+    SECTORS=$(blockdev --getsz "$DEVICE")
+    # Unmount possible mounted filesystems
+    sync; umount -q "$DEVICE"* || true;
+    # Wipe first 10MiB of disk
```
Comment (Mic92) [inline packages/flash/flash.sh:52]: Good catch, we should add this to disko as well.

File: docs/src/ref_impl/build_and_run.md [hunk 4]
```diff
@@ -192,11 +192,11 @@ In the case of the Icicle Kit, Ghaf deployment consists of creating an SD image
 2. Flash the Ghaf SD image:
 
    * If you want to use a SD card:
-     * Prepare the SD card with the target HW image you built: `dd if=./result/nixos.img of=/dev/<YOUR_SD_DEVICE> bs=32M status=progress oflag=direct`.
+     * Prepare the SD card with the target HW image you built: `./packages/flash/flash.sh -d /dev/<YOUR_SD_CARD> -i result/<IMAGE_NAME>`.
      * Insert an SD card into the board and switch the power on.
 
    * If you want to use the onboard MMC:
-     * You can directly flash a NixOS image to onboard an MMC card: `dd if=./result/nixos.img of=/dev/<YOUR_MMC_DEVICE> bs=32M status=progress oflag=direct`.
+     * You can directly flash a NixOS image to an onboard MMC card: `dd if=./result/nixos.img of=/dev/<YOUR_MMC_DEVICE> bs=32M status=progress oflag=direct`.
```
[the remark this answers is not in the document]
Comment (unbel13ver) [reply docs/src/ref_impl/build_and_run.md]: I did not know that, thanks. Addressed.

File: docs/src/ref_impl/build_and_run.md [hunk 5]
```diff
@@ -128,9 +128,9 @@ After the latest firmware is [flashed](./build_and_run.md#flashing-nvidia-jetson
     ```
     nix build github:tiiuae/ghaf#nvidia-jetson-orin-agx-debug
     ```
-2. After the build is completed, prepare a USB boot media with the target image you built:
+2. After the build is completed, prepare a USB boot media with the target image you built using the `flash.sh` script:
     ```
-    dd if=./result/nixos.img of=/dev/<YOUR_USB_DRIVE> bs=32M status=progress oflag=direct
+    ./packages/flash/flash.sh -d /dev/<YOUR_USB_DRIVE> -i result/<IMAGE_NAME>
```
[the remark this answers is not in the document]
Comment (unbel13ver) [reply docs/src/ref_impl/build_and_run.md]: Addressed, thanks.

**Review by brianmcgillion — APPROVED** [review]
LGTM - improvements can come later, if needed

**Pull request merged**